In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv("https://catalog.ourworldindata.org/garden/covid/latest/compact/compact.csv")

In [3]:
df

,country,date,total_cases,new_cases,new_cases_smoothed,total_cases_per_million,new_cases_per_million,new_cases_smoothed_per_million,total_deaths,new_deaths,...,population,population_density,median_age,life_expectancy,gdp_per_capita,extreme_poverty,diabetes_prevalence,handwashing_facilities,hospital_beds_per_thousand,human_development_index
0,Afghanistan,2020-01-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
1,Afghanistan,2020-01-02,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
2,Afghanistan,2020-01-03,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
3,Afghanistan,2020-01-04,0.0,0.0,NaN,0.000,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
4,Afghanistan,2020-01-05,0.0,0.0,NaN,0.000,0.0,NaN,0.0,0.0,...,40578847.0,62.215549,16.752001,65.616997,1983.812622,NaN,11.7,51.938343,0.35,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
591517,Zimbabwe,2026-05-13,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
591518,Zimbabwe,2026-05-14,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
591519,Zimbabwe,2026-05-15,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN
591520,Zimbabwe,2026-05-16,266436.0,0.0,0.0,16580.682,0.0,0.0,5740.0,0.0,...,16069061.0,41.538223,17.683001,62.360100,5215.252930,49.219894,1.5,42.509991,1.95,NaN


# **Basic Exploratory Data Analysis**

## <u>*Dataset Info*</u> 

In [18]:
df.shape

(591522, 61)

In [19]:
df.sample(5)

,country,date,total_cases,new_cases,new_cases_smoothed,total_cases_per_million,new_cases_per_million,new_cases_smoothed_per_million,total_deaths,new_deaths,...,population,population_density,median_age,life_expectancy,gdp_per_capita,extreme_poverty,diabetes_prevalence,handwashing_facilities,hospital_beds_per_thousand,human_development_index
204440,Greenland,2024-02-01,11971.0,0.0,0.000000,213714.420000,0.000000,0.000000,21.0,0.0,...,56014.0,0.136470,33.924000,70.051903,71037.539062,NaN,4.3,NaN,NaN,NaN
85444,Burundi,2024-06-15,54569.0,0.0,0.000000,4096.433000,0.000000,0.000000,15.0,0.0,...,13321102.0,518.734497,15.672000,62.881901,1051.259277,74.199173,4.3,6.359708,0.69,NaN
35581,Austria,2021-10-29,805051.0,4302.0,3677.714400,88811.870000,474.589400,405.719240,13798.0,30.0,...,9064678.0,109.848251,42.832001,81.295898,63788.078125,0.528223,5.4,NaN,6.70,NaN
302478,Luxembourg,2020-03-08,4.0,3.0,0.428571,6.122621,4.591966,0.655995,0.0,0.0,...,653315.0,253.767776,38.675999,82.200699,128475.281250,0.085138,5.8,NaN,3.95,NaN
202887,Greece,2026-03-18,5864139.0,0.0,100.857140,563183.600000,0.000000,9.686178,40246.0,0.0,...,10412481.0,80.779526,45.390999,79.944099,37474.281250,0.594283,8.0,NaN,4.27,NaN


In [6]:
df.info()
df["date"]=pd.to_datetime(df["date"]) 
# converting date from object to datetime

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 591522 entries, 0 to 591521
Data columns (total 61 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   country                                     591522 non-null  object 
 1   date                                        591522 non-null  object 
 2   total_cases                                 579174 non-null  float64
 3   new_cases                                   574898 non-null  float64
 4   new_cases_smoothed                          573689 non-null  float64
 5   total_cases_per_million                     579174 non-null  float64
 6   new_cases_per_million                       574898 non-null  float64
 7   new_cases_smoothed_per_million              573689 non-null  float64
 8   total_deaths                                579174 non-null  float64
 9   new_deaths                                  576007 non-null  float64
 

In [15]:
df.isnull().sum().sort_values(ascending=False)

human_development_index                    591522
weekly_icu_admissions                      580529
weekly_icu_admissions_per_million          580529
excess_mortality_cumulative_per_million    577749
excess_mortality                           577715
                                            ...  
total_cases_per_million                     12348
total_deaths                                12348
total_deaths_per_million                    12348
date                                            0
country                                         0
Length: 61, dtype: int64

***Many column have missing value more than 50% of total data***

In [16]:
size=np.round(df.memory_usage(deep=True).sum()/1024**2,2)
print(f"DataFrame(df) size : {size} MB")

DataFrame(df) size : 354.66 MB


In [17]:
df.duplicated().sum()

np.int64(0)

- *Observation*<br>
    - The shape of DataFrame is (589779, 61).<br>
    - Many column have missing value more than 50% of total data.<br>
    - Size of DataFrame :382.3 MB
***

***
## <u>Columns Analysis</u>

- **Objective**<br>
    - *Checking each Column thorougly and getting important information out from it.*<br>
    - *Get important Pattern out from different columns.*<br>

### Country AND Date Analysis

In [20]:
countries=df["country"].unique()
no_of_unique_country=df["country"].nunique()
no_of_unique_country

262

In [23]:
print(f"Maximum days data: {df["country"].value_counts().max()}")
print(f"Minimum days data: {df["country"].value_counts().min()}")
print("for each country")

Maximum days data: 2329
Minimum days data: 1
for each country


In [57]:
country_count=df["country"].value_counts()
Countries=country_count[country_count==2329].index
df_complete=df[df["country"].isin(Countries)]
df_complete["country"].value_counts()
Countries

Index(['Zimbabwe', 'Afghanistan', 'Albania', 'Algeria', 'Uruguay',
       'United States Virgin Islands', 'United Arab Emirates',
       'United Kingdom', 'United States', 'Cambodia',
       ...
       'Saudi Arabia', 'Seychelles', 'Slovakia', 'Slovenia', 'Sierra Leone',
       'Romania', 'Portugal', 'Puerto Rico', 'Qatar', 'South Africa'],
      dtype='object', name='country', length=181)

In [13]:
missing_count=df.groupby("country")["total_cases"].apply(lambda x : x.isnull().sum())
missing_count.sort_values(ascending=False).head(5)

country
Hong Kong           1672
Taiwan              1386
Northern Ireland    1385
England             1359
Scotland            1327
Name: total_cases, dtype: int64

-  *Observation*
    - The maximum number of missing values in `total_cases` is observed for Hong Kong (1672 missing records).
    - Several countries contain substantial missing data and may require filtering before comparative analysis.
    - 181 countries out of total 262 countries(Involving Olympic Event And World Data) have complete records covering all 2322 days.

In [50]:
last_date_of_record=df["date"].max()   
first_date_of_record=df["date"].min()

In [59]:
country_date_range = (
    df.groupby("country")["date"]
      .agg(
          start_date="min",
          end_date="max",
          count="count"
      )
      .reset_index()
)

country_date_range

,country,start_date,end_date,count
0,Afghanistan,2020-01-01,2026-05-17,2329
1,Africa,2020-01-04,2026-05-17,2326
2,Albania,2020-01-01,2026-05-17,2329
3,Algeria,2020-01-01,2026-05-17,2329
4,American Samoa,2020-01-04,2026-05-17,2326
...,...,...,...,...
257,World excl. China and South Korea,2020-01-04,2026-05-17,2326
258,"World excl. China, South Korea, Japan and Sing...",2020-01-04,2026-05-17,2326
259,Yemen,2020-01-01,2026-05-17,2329
260,Zambia,2020-01-01,2026-05-17,2329


In [60]:
(last_date_of_record-first_date_of_record).days+1

2329

- *Observation*
    - Most countries have records from 2020-01-01 to 2026-05-10.
    - Some countries/entities started reporting later (e.g., Africa starts on 2020-01-04).
    - Countries with fewer records may indicate missing or incomplete reporting.

### Case Analysis

In [100]:
total_cases_max = df.groupby("country")["total_cases"].max()

total_cases_max.sort_values(ascending=False).head(15)
# total_cases_max["India"]
Cases_per_country=(df.groupby("country").agg(
    Total_Cases = ("total_cases","max"),
    Max_New_Cases= ("new_cases","max")
).reset_index())
peak_dates=df.loc[
    df.groupby("country")["new_cases"].idxmax().dropna(), 
    ["country","date"]
]
peak_dates=peak_dates.rename( columns={"date":"max_newcase_date"})
# used dropna bcz peak_dates was getting some NAN values
Cases_per_country=Cases_per_country.merge(peak_dates,on="country")
# Cases_per_country[Cases_per_country["country"]=="India"]
Cases_per_country.sort_values(by="Max_New_Cases",ascending=False).head(10)

,country,Total_Cases,Max_New_Cases,max_newcase_date
242,World,779135536.0,8401906.0,2022-01-30
243,World excl. China,679753775.0,8401713.0,2022-01-30
244,World excl. China and South Korea,645181902.0,8384638.0,2022-01-30
245,"World excl. China, South Korea, Japan and Sing...",608372175.0,8295098.0,2022-01-30
12,Asia,302181905.0,7213802.0,2022-12-23
96,High-income countries,458049922.0,7197333.0,2022-01-30
234,Upper-middle-income countries,239832298.0,7027032.0,2022-12-23
45,China,99381761.0,6966046.0,2022-12-23
70,Europe,254960735.0,6386449.0,2022-01-30
71,European Union (27),186985838.0,5896096.0,2022-01-30


In [101]:
Cases_per_country[Cases_per_country["country"].isin(["China","India"])]

,country,Total_Cases,Max_New_Cases,max_newcase_date
45,China,99381761.0,6966046.0,2022-12-23
100,India,45056221.0,414188.0,2021-05-07


In [102]:
country_cases=Cases_per_country[~Cases_per_country["country"].str.contains("World|excl.|Europe|Asia|North|South|income|Upper|European")]
country_cases.sort_values(by="Total_Cases",ascending=False)


,country,Total_Cases,Max_New_Cases,max_newcase_date
232,United States,103436829.0,1265520.0,2022-01-12
45,China,99381761.0,6966046.0,2022-12-23
100,India,45056221.0,414188.0,2021-05-07
76,France,39060446.0,2417043.0,2022-01-23
82,Germany,38437953.0,1588891.0,2022-03-27
...,...,...,...,...
159,Niue,1092.0,117.0,2023-01-02
220,Tokelau,80.0,39.0,2023-07-15
238,Vatican,26.0,7.0,2020-10-13
175,Pitcairn,4.0,4.0,2022-07-20


In [104]:
print(df["new_cases_smoothed"].max())
print(df["new_cases"].max())
df[["date","new_cases","new_cases_smoothed"]].sort_values(by="new_cases",ascending=False).head(8)

6402033.0
8401906.0


,date,new_cases,new_cases_smoothed
575988,2022-01-30,8401906.0,3346287.8
578314,2022-01-30,8401713.0,3346053.2
580640,2022-01-30,8384638.0,3331154.5
575981,2022-01-23,8346273.0,3362941.2
578307,2022-01-23,8345987.0,3362752.2
580633,2022-01-23,8338479.0,3356269.5
582966,2022-01-30,8295098.0,3254791.2
582959,2022-01-23,8286572.0,3311735.8


- *Obsevation*
    - Highest case in a day reported by any Individual country is China and which is On 2022-12-2023
    - Total cases in India was 45056221 and maximum new_cases found in a day in India was 414188 On 2021-05-07
    - In dataset(df) there is a data of different continents which will always be more then individual countries so for different country comparison we need to filter out countries.
    - United States had highest number of cases followed by China and India.(It shows that USA,China and India were most affected)
    - Smoothed features shows the seven days cummulative avg of all the new cases such that it become smoothed new cases.Because Covid do not exploded suddenly so we can get better trend analysis with this smoothed data

### Death Analysis

In [138]:
df_deaths=df.groupby("country").agg(
    total_deaths=("total_deaths","max"),
    new_deaths_max=("new_deaths","max")
).reset_index().sort_values(by="new_deaths_max",ascending=False)

peak_death_date=df.loc[df.groupby("country")["new_deaths"].idxmax().dropna(),("country","date")]
df_deaths=df_deaths.merge(peak_death_date,on="country")
df_deaths

df_deaths=df_deaths[~df_deaths["country"].str.contains("World|excl.|Europe|Asia|North|South|income|Upper|European")]

df_deaths
# we will merge this table with last country_cases table to get complete analysis of cases and death all together and extract important information out
df_final_case_death=pd.merge(country_cases,df_deaths,on="country",how="left")
df_final_case_death["case_fatality_rate"]=np.round((df_final_case_death["total_deaths"]*100/df_final_case_death["Total_Cases"]),4)
df_final_case_death.sort_values(by="Total_Cases",ascending=False)
df_final_case_death.sort_values(by="new_deaths_max",ascending=False).head(10)

,country,Total_Cases,Max_New_Cases,max_newcase_date,total_deaths,new_deaths_max,date,case_fatality_rate
43,China,99381761.0,6966046.0,2022-12-23,122398.0,44047.0,2023-02-01,0.1232
42,Chile,5410524.0,41651.0,2022-02-12,64497.0,11447.0,2022-03-22,1.1921
104,Japan,33803572.0,326090.0,2022-08-29,74694.0,9903.0,2023-02-01,0.2210
60,Ecuador,1082192.0,11536.0,2020-04-25,36064.0,8786.0,2021-07-21,3.3325
78,Germany,38437953.0,1588891.0,2022-03-27,174979.0,6460.0,2020-12-20,0.4552
95,India,45056221.0,414188.0,2021-05-07,533849.0,6148.0,2021-06-10,1.1849
194,Spain,13980340.0,956506.0,2022-01-16,121886.0,5841.0,2020-04-05,0.8718
72,France,39060446.0,2417043.0,2022-01-23,168207.0,5602.0,2020-11-22,0.4306
217,United States,103436829.0,1265520.0,2022-01-12,1238189.0,5061.0,2021-02-14,1.1970
214,Ukraine,5555510.0,242942.0,2022-02-06,109937.0,4591.0,2021-11-14,1.9789


In [140]:
df_final_case_death.sort_values(by="case_fatality_rate",ascending=False).head(10)

,country,Total_Cases,Max_New_Cases,max_newcase_date,total_deaths,new_deaths_max,date,case_fatality_rate
226,Yemen,11945.0,287.0,2022-02-08,2159.0,60.0,2022-02-08,18.0745
196,Sudan,63993.0,1284.0,2022-02-20,5046.0,951.0,2022-03-04,7.8852
200,Syria,57423.0,728.0,2021-10-21,3163.0,21.0,2021-10-19,5.5082
193,Somalia,27334.0,1066.0,2022-01-30,1361.0,86.0,2021-11-05,4.9791
162,Peru,4533813.0,59470.0,2022-01-23,221071.0,843.0,2021-04-23,4.8761
61,Egypt,516023.0,5516.0,2022-03-19,24830.0,97.0,2020-06-16,4.8118
133,Mexico,7630026.0,81302.0,2022-01-19,335106.0,1495.0,2021-01-22,4.3919
27,Bosnia and Herzegovina,404289.0,3342.0,2022-01-19,16419.0,99.0,2021-03-30,4.0612
117,Liberia,8090.0,219.0,2021-07-03,294.0,14.0,2020-07-14,3.6341
0,Afghanistan,235214.0,3243.0,2021-06-17,7998.0,159.0,2021-06-17,3.4003


- **Observations**

    * Countries with a larger number of cumulative COVID-19 cases did not always experience the highest daily death counts. This suggests that total infections alone cannot fully explain the severity of mortality across countries.
    
    * In most countries, the peak in daily deaths occurred after the peak in daily cases. This time gap indicates that the impact of a surge in infections is often reflected in mortality statistics after a delay.
    
    * China recorded the highest single-day death count in the dataset, reporting **44,047 deaths on 2023-02-01**. Other countries with notably high peak daily deaths include **Chile, Japan, Ecuador, Germany, and India**.
    
    * India reported its highest daily death count of **6,148 deaths on 2021-06-10**, which coincides with the period following the severe second wave of COVID-19 in the country.
    
    * The dates of maximum daily cases and maximum daily deaths are generally different, highlighting that infection waves and mortality waves do not peak simultaneously.
    
    * Peak daily death counts provide useful insight into the intensity and impact of COVID-19 waves experienced by different countries during the pandemic.
    
    * Overall, the analysis shows that while case counts are an important indicator of disease spread, mortality trends are influenced by multiple factors and may follow a different pattern over time.
 
    * **Yemen** recorded the highest case fatality rate in the dataset (18.07%), making it a clear outlier compared to other countries. This may reflect factors such as limited testing capacity, under-reporting of infections, healthcare challenges, or differences in reporting practices.


### Hospital_Analysis

In [146]:
df[['hosp_patients',
       'hosp_patients_per_million', 'weekly_hosp_admissions',
       'weekly_hosp_admissions_per_million', 'icu_patients',
       'icu_patients_per_million', 'weekly_icu_admissions',
       'weekly_icu_admissions_per_million']].isnull().mean().sort_values(ascending=False)*100

weekly_icu_admissions                 98.141574
weekly_icu_admissions_per_million     98.141574
weekly_hosp_admissions_per_million    95.858649
weekly_hosp_admissions                95.858649
icu_patients_per_million              93.387228
icu_patients                          93.387228
hosp_patients                         93.126883
hosp_patients_per_million             93.126883
dtype: float64

- *Observation*

    - Hospitalization and ICU-related variables contain extremely high missing values (above 93% across all hospital metrics).
    
    - Due to the limited data availability, a meaningful comparison of hospitalization trends across all countries would not be statistically reliable.
    
    - Therefore, hospital-related features will not be included in the global comparative analysis and the focus will be shifted to other variables with better data coverage.



### Other

In [196]:

df[['stringency_index',
       'reproduction_rate', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case']].isnull().mean()

stringency_index                   0.657223
reproduction_rate                  0.686093
total_tests                        0.865792
new_tests                          0.872527
total_tests_per_thousand           0.865792
new_tests_per_thousand             0.872527
new_tests_smoothed                 0.824242
new_tests_smoothed_per_thousand    0.824242
positive_rate                      0.828307
tests_per_case                     0.829846
dtype: float64

In [197]:
df.groupby("country")["total_tests"].max().notnull().sum()

np.int64(188)

In [198]:
# df_others=df.groupby("country").agg(
#     stringency_max=("stringency_index","max")
# ).reset_index()
# df_others.sort_values(by="stringency_max",ascending=False).head(20)
peak_stringency=df.loc[df.groupby("country")["stringency_index"].idxmax().dropna(),("country","stringency_index","date")].sort_values(by="stringency_index",ascending=False).head(10)
peak_stringency

,country,stringency_index,date
21031,Argentina,100.0,2020-03-23
150290,El Salvador,100.0,2020-05-07
122349,Cuba,100.0,2020-05-11
140965,Dominican Republic,100.0,2020-04-28
290879,Libya,100.0,2020-04-17
228632,Honduras,100.0,2020-03-21
237292,India,100.0,2020-03-22
262901,Jordan,100.0,2020-03-18
191396,Georgia,100.0,2020-03-31
393667,Oman,100.0,2020-07-25


In [199]:
df[["stringency_index","reproduction_rate"]].corr()

,stringency_index,reproduction_rate
stringency_index,1.000000,0.291778
reproduction_rate,0.291778,1.000000


- *Observations*
    
    - Several countries, including China, the United States, and Japan, contain missing values in the stringency index, which may limit direct policy comparisons.
    
    - Multiple countries reached the maximum stringency index value of 100 during the pandemic, indicating the implementation of extremely strict containment measures such as lockdowns, school closures, workplace closures, travel restrictions, and stay-at-home orders.
    
    - India reached the maximum stringency index of 100 on 2020-03-22, reflecting one of the strictest phases of pandemic control measures in the country.
    
    - Countries achieved their peak stringency levels at different points in time, suggesting that governments responded to local outbreak conditions and policy priorities differently.
 


### Test Analysis

In [201]:
df.columns

Index(['country', 'date', 'total_cases', 'new_cases', 'new_cases_smoothed',
       'total_cases_per_million', 'new_cases_per_million',
       'new_cases_smoothed_per_million', 'total_deaths', 'new_deaths',
       'new_deaths_smoothed', 'total_deaths_per_million',
       'new_deaths_per_million', 'new_deaths_smoothed_per_million',
       'excess_mortality', 'excess_mortality_cumulative',
       'excess_mortality_cumulative_absolute',
       'excess_mortality_cumulative_per_million', 'hosp_patients',
       'hosp_patients_per_million', 'weekly_hosp_admissions',
       'weekly_hosp_admissions_per_million', 'icu_patients',
       'icu_patients_per_million', 'weekly_icu_admissions',
       'weekly_icu_admissions_per_million', 'stringency_index',
       'reproduction_rate', 'total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case', 'total_vaccinations',
    

In [204]:
df[['total_tests', 'new_tests',
       'total_tests_per_thousand', 'new_tests_per_thousand',
       'new_tests_smoothed', 'new_tests_smoothed_per_thousand',
       'positive_rate', 'tests_per_case']].isnull().mean()

total_tests                        0.865792
new_tests                          0.872527
total_tests_per_thousand           0.865792
new_tests_per_thousand             0.872527
new_tests_smoothed                 0.824242
new_tests_smoothed_per_thousand    0.824242
positive_rate                      0.828307
tests_per_case                     0.829846
dtype: float64

In [210]:
df.groupby("country")["total_tests"].max().notnull().sum()

np.int64(188)

In [254]:
df_test=(df.groupby("country").agg(
    total_test=("total_tests","max"),
    total_test_per_thousand=("total_tests_per_thousand","max"),
    max_positive_rate=("positive_rate","max"),
    avg_positive_rate=("positive_rate","mean"),
    avg_tests_per_case=("tests_per_case","mean")
).reset_index())

df_test[df_test["country"]=="India"]

peak_positive_rate=df.loc[df.groupby("country")["positive_rate"].idxmax().dropna(),("country","positive_rate","date")]
peak_positive_rate[peak_positive_rate["country"]=="India"]

,country,positive_rate,date
237707,India,22.298986,2021-05-11


In [255]:
df_test.sort_values(by="avg_positive_rate",ascending=False).head(5)

,country,total_test,total_test_per_thousand,max_positive_rate,avg_positive_rate,avg_tests_per_case
220,Suriname,84184.0,137.335,94.374650,51.262347,2.582644
31,Brazil,70923215.0,330.912,94.406784,46.827564,3.307875
54,Curacao,22100.0,116.109,94.505490,46.212448,1.537818
145,Mexico,15569464.0,122.879,76.051960,30.618106,19.698492
41,Cayman Islands,9358.0,137.343,33.888920,30.486465,3.139729


In [256]:
df_test.sort_values(by="avg_tests_per_case",ascending=False).head(5)

,country,total_test,total_test_per_thousand,max_positive_rate,avg_positive_rate,avg_tests_per_case
45,China,9.214000e+09,6461.913,0.275661,0.009187,262255.629445
47,Comoros,1.002890e+05,122.062,0.229709,0.067668,3291.203421
205,Sierra Leone,4.285170e+05,50.889,0.239202,0.039192,2571.975407
26,Bhutan,1.711669e+06,2201.543,4.076423,0.369677,2282.385206
14,Australia,7.337030e+07,2830.525,55.162525,8.774566,1739.672971


In [257]:
df_test.sort_values(by="total_test_per_thousand",ascending=False).head(5)

,country,total_test,total_test_per_thousand,max_positive_rate,avg_positive_rate,avg_tests_per_case
55,Cyprus,29501771.0,32925.824,5.801486,1.366734,254.029574
15,Austria,189791719.0,21272.133,38.945858,3.560339,264.663919
240,United Arab Emirates,168522672.0,17994.670,2.291004,0.706391,601.154932
87,Gibraltar,533268.0,16322.865,11.030626,8.499577,12.580708
76,Faroe Islands,777845.0,14707.401,63.542603,25.608018,12.229658


In [259]:
df_test.sort_values(by="total_test",ascending=False).head(5)

,country,total_test,total_test_per_thousand,max_positive_rate,avg_positive_rate,avg_tests_per_case
45,China,9.214000e+09,6461.913,0.275661,0.009187,262255.629445
242,United States,9.127691e+08,2708.533,33.763290,8.311368,16.879474
104,India,8.585264e+08,609.938,22.298986,4.705920,101.019837
241,United Kingdom,5.032703e+08,7480.121,28.009380,4.788321,74.090877
188,Russia,2.952448e+08,2034.729,29.323301,5.281616,68.164916


In [262]:
df_test[["avg_positive_rate",
         "avg_tests_per_case"]].corr()

,avg_positive_rate,avg_tests_per_case
avg_positive_rate,1.000000,-0.098228
avg_tests_per_case,-0.098228,1.000000


In [263]:
df_test[["avg_positive_rate",
         "total_test_per_thousand"]].corr()

,avg_positive_rate,total_test_per_thousand
avg_positive_rate,1.000000,-0.108104
total_test_per_thousand,-0.108104,1.000000


- *Observations*

    * Testing data was available for 188 countries, which provided sufficient coverage for performing a meaningful global testing analysis despite the presence of missing values in some testing-related variables.
    
    * China conducted the highest number of COVID-19 tests, followed by the United States, India, the United Kingdom, and Russia. This trend is expected since countries with larger populations generally require a greater number of tests to monitor and control disease spread.
    
    * When testing figures were adjusted for population size, a different pattern emerged. Countries such as Cyprus, Austria, the United Arab Emirates, Gibraltar, and the Faroe Islands recorded the highest number of tests per thousand people, indicating a much more intensive testing strategy relative to their population.
    
    * Suriname, Brazil, and Curacao showed the highest average positivity rates in the dataset. A high positivity rate often suggests that infections were spreading rapidly or that testing capacity was not sufficient to capture the true extent of infections within the population.
    
    * China recorded the highest tests-per-case ratio, followed by Comoros, Sierra Leone, Bhutan, and Australia. This indicates that these countries performed a large number of tests for every confirmed case detected, reflecting extensive surveillance and proactive testing efforts.
    
    * A clear difference can be observed between countries with high testing intensity and those with high positivity rates. Countries that conducted widespread testing generally reported lower positivity rates, while countries with limited testing coverage often experienced higher positivity rates.
    
    * Correlation analysis between average positivity rate and testing intensity revealed only a very weak negative relationship. This suggests that while increased testing may contribute to lower positivity rates, other factors such as outbreak severity, population characteristics, government policies, and reporting practices also played an important role.
    
    * In India, the highest positivity rate (22.3%) was recorded on 11 May 2021, only a few days after the peak in daily new cases on 7 May 2021. This coincides with the country's severe second wave and reflects the period of intense community transmission.
